### ISSUE TITLE
Consider getting rid of function codes.

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#1415]
This is no doubt too late for 1.1.0, but perhaps 1.2.0. (Or perhaps earlier if you don't mind making some APIs start kinda lying.)

The function codes in err are kind of nuts. They make development cumbersome (function codes must be updated whenever a new function that uses err is added), are a backwards compatibility nightmare if anyone tries to condition on them (rare, but happens on occasion), easy to mess up (code routinely uses the wrong one?), and impede refactoring like renaming non-public functions.

In BoringSSL, we got rid of them. First with the following scheme:
- The error code itself never contains a function code number. It's always zero.
- `ERR_func_error_string` always returns the special value `OPENSSL_internal`. Imagine that OpenSSL decided to turn every function into a wrapper over a static function named `OPENSSL_internal`. It'd be nonsense but legal C. This extends to be behavior of `ERR_error_string_n`.
- `ERR_put_error` currently samples and stores `__FILE__` and `__LINE__` out-of-band. Also sample `__func__` and store that. This requires changing `ERR_put_error` or adding an `ERR_put_error_ex`.
- `ERR_print_errors_cb` extracted this internal value and, rather than the unhelpful `OPENSSL_internal`, printed the true out-of-band function name.

That was a huge productivity win, but not a binary size one. We later decided sampling `__func__` really wasn't very useful and wasted binary size in every BoringSSL consumer. We undid the changes to `ERR_put_error` and `ERR_print_errors_cb`. Now we always report function names as `OPENSSL_internal`.

This is fairly simple and seems to have worked just fine. If you're using the function codes for logging, the file and line number are readily available and strictly more useful. If you're using them for anything programmatic... don't. No other library lets consumers condition on the internal implementation details like this.

**Comment by levitte** [comment]
This is definitely post 1.1.0, yes...  and 1.2.0 sounds right to me.

That being said, there are a few of us in the team that have been talking about remaking the whole error code stuff, so we'll see things happening

**Comment by davidben** [comment]
Is there a summary somewhere of your plans in this space? I definitely have my own laundry list of problems with the current error system. I'm curious how many of them overlap.

**Comment by richsalz** [comment]
no summary.  this is from some internal team email i wrote that started the discussion
I was thinking about the error string tables.

; find . -name '*_err.o' | xargs size | awk 'NR>1{X+=$4;} END {print X;}'
87224

It would be a win if we could move that to text (const) space, and avoid some initialization time.  Here's a suggestion on how to do that.  Put all error strings for libcrypto into one file that has the table sorted.  Put all error strings for libssl into one file (well, it already is) that also has the table sorted.  The "global error table" becomes a linked list of pointers to error tables, perhaps with min/max error codes.  Adding a new error table dynamically (such as engines do, or as libssl does) would just add an entry to that linked list.  Something like this:

typedef struct error_table {
	const ERR_STRING_DATA *table;
	size_t tsize;
	unsigned log tmin, tmax;
	struct error_table *next;
} ERROR_TABLE;

Searching an error code walks the linked list, when it finds a possible error table, it does a bsearch.  In almost all cases it will find it in either libcrypto or libssl, the first or second table in the list.  All of the individual ERR_load_xxxx_strings go away.

Most of the change would be in writing a new mkerr script.  If that script can't be changed to create the table in-order, then we add to the ERROR_TABLE an array of pointers to ERR_STRING_DATA that are sorted. I'd rather not do that, and would instead teack mkerr about the #define's so it can do the PACK_ERR stuff itself and just generate the hex codes directly. 

Also, it should put the codes into separate files, not rewrite the existing headers.  Ugh.

**Comment by richsalz** [comment]
As discussed before, openssl has the requirement that third-parties can add error tables; boring doesn't.

**Comment by richsalz** [comment]
Previous team discussions did NOT want to lose the function name.  I do.  Maybe attitudes will change.  I believe the file/line number is more than sufficient.

**Comment by davidben** [comment]
I don't see what adding new error tables has to do with removing function codes. That's equally doable either way.

And yeah, you should make it a const table too. We've done that and it makes things dramatically nicer. It's certainly easier when you don't allow custom error strings, but still perfectly doable. Just treat them differently. It's easy to distinguish by lib number.

I was wondering about any planned semantic changes, like this suggestion or others. There's a ton of problems with err, though fixing some may involve breakage.

**Comment by richsalz** [comment]
You asked what we’re looking at doing, I told you some things.  I'd like to know what tons of problems you see?

**Comment by davidben** [comment]
Ah, I see. Sorry, I misunderstood Richard and thought he meant you all were planning an overhaul of the semantics and not just tightening the internals. When the more thorough overhaul is on the radar, here are some things I've noticed from dealing with existing OpenSSL consumers:

* Function codes are awkward. That's this bug.
* Reason codes are very specific. It's a bit tricky because you have these competing goals where you want detailed logging for diagnostics, but you also want sufficiently high-level error codes so that, when consumers do need to condition on particular ones, they can do so without imposing compatibility constraints. The current error system works well for the first use and very badly for the second. We've repeatedly run into issues where, as a result of cleaning up code, the old error codes don't apply anymore and existing code gets screwy.
* The two-level namespacing is confusing. I've seen a lot of code which conditions on the reason code without checking the library code. I've been thinking of hacking up our stack so `ERR_GET_REASON` and `SSL_R_FOO_BAR` values both keep the library code OR'd in. We'll see if that works.
* The two-level namespacing leaks internal abstraction boundaries. We've run into issues where we've moved code around in ways invisible to public APIs, except that error codes now appear in a different library. We've ended up defining "cross-module errors" to work around this. (By the way, cross-module errors are awkward if you don't ditch function codes. Doable, but messy.)
* Code within the library is very inconsistent about to and not to push a "stack trace" error.
* Consumers programmatically consuming error codes tend not to know what to do with the queue.
* No one remembers to call `ERR_clear_error`.

That last one needs expanding. With `errno`, one rarely needs to clear it. If I call, say, `write` with `errno` state from a previous operation lying around, everything's fine. Whether the operation succeeded or failed is determined by the return code. If the operation succeeded, it doesn't touch `errno`, but I don't not to check it. If it failed, the operation always overrides `errno`, so the previous state doesn't collide.

With the error queue, neither is true. First, `SSL_get_error` behaves differently based on whether the error queue is empty or not. Which means your SSL code will get confused if you leave errors in the queue. We fixed this by calling `ERR_clear_error` at the very top of everything API paired with `SSL_get_error`. Second, in the failure case, the error queue ends up with errors from both operations munged together. This is screwy. I haven't found a very easy way to fix this while retaining the queue portion. Some further thoughts [here](https://bugs.chromium.org/p/boringssl/issues/detail?id=38).

**Comment by levitte** [comment]
Actually, there's no saying what we'll end up doing, so semantics may change.  That's why I see this primarly as a 1.2.0 activity.

**Comment by levitte** [comment]
I just went reading the further thoughts, and they were similar to thoughts I had around this quite a while ago (never finished, this was more than 10 years ago, and I got distracted by other things at the time)...

The thing I got stuck with every time was figuring out where to start a new queue in a safe way.  Considering that our public API isn't just for external callers, but are called by ourselves as well, it seemed impossible to know beforehand where the boundary between the external caller and ourselves would be in reality.

A quick thought now is that we could have some kind of (per thread) call depth counter that gets incremented upon entry of any of our functions and decremented upon exit from the same.  It being zero at the time of incrementing would automatically clear the error stack for the current thread.
This would mean that *every* function needs some `enter` and `leave` code that does this checking for zero, incrementing and decrementing...  quite simple to do with three macros (ENTER, LEAVE and RETURN), tedious to insert the invocations everywhere.

**Comment by davidben** [comment]
If we're fine inserting and removing code into every function, I think `ERR_clear_error` at the start of every function would work about as well. But yeah, it'd be quite tedious. I was hoping to avoid that kind of thing.

The best idea I've had so far (though it too is a little awkward) was to split `ERR_put_error` in two: `ERR_start_error` and `ERR_add_to_error`. The former means throw a new exception and is when we start an error. It implicitly clears everything. The latter means to append to the existing error.

That, however, requires the library be much much more thorough about errors. If every function which failed *always* added an error to the error queue on every failure path, then this convention works out. Otherwise it becomes awkward:
https://bugs.chromium.org/p/boringssl/issues/detail?id=38#c2

(I may give this idea a try when I find the cycles. I'll certainly let you all know if it works out.)

**Comment by richsalz** [comment]
Inserting code into every function seems very bad. The semantics of "add error" and "set error" and maybe "add detail to stack" seem reasonable.

**Comment by levitte** [comment]
Wellll....  Just as a quick example; there's the function `PEM_read_bio_X509`, callable by anyone, and *also* called from within the libssl code (among others).  How does `PEM_read_bio_X509` know if it should *start* an error queue or *add* to it?

**Comment by richsalz** [comment]
Something for our F2F tech discussion.

**Comment by davidben** [comment]
@levitte: If the error originates from `PEM_read_bio_X509`, it starts an error queue. When the libssl code wants to push an `ERR_R_PEM_LIB` on top of it, after `PEM_read_bio_X509` has failed, it adds to it.

The difficulty comes from `PEM_read_bio_X509` having some codepaths which do touch the error queue and some which don't. Then libssl doesn't know whether to add or start. Ideally we just wouldn't have those but `BIO` in particular is a mess here, with the retry "errors" which signal failure but don't touch the error queue, the possibility of external callbacks, etc. I suspect there will need to be defensive `ERR_clear_error` calls in a few places. But hopefully they can be in targeted places where legacy API patterns dictate it, rather than everywhere.

**Comment by levitte** [comment]
Ah, yes, I get what you're saying.

**Comment by richsalz** [comment]
Also, I strongly think error-code/filename/line# is really all that’s needed to debug.

**Comment by levitte** [comment]
After much thought, I agree...  *as loong as* the error code includes both a library module number and a reason code.  So in some future, it sounds like we're munging it down to a call like this:

    FOOerr(FOO_R_whatever);

(that's mimicing the current syntax);

**Comment by richsalz** [comment]
Why do you want a module?  File/line# seems sufficient.

**Comment by davidben** [comment]
Right now the error values are a two-level namespace, so you need both the module and the reason code. The reason codes have the module name in them, the values don't stand alone. Ours is similar. We went from `OPENSSL_PUT_ERROR(FOO, FOO_do_something, FOO_R_REASON)` to `OPENSSL_PUT_ERROR(FOO, FOO_R_REASON)` when ditching function codes. See the third bullet point in my list.

(Though I suspect this could be changed while maintaining source compatibility, though not binary compatibility... I'll give that a try sometime. Cutting it down to `OPENSSL_PUT_ERROR(FOO_R_REASON)` would be fun.)

**Comment by richsalz** [comment]
There are two schools of thought, either error codes are a global flat namespace (errno), or each error code can be used uniquely (most IBM stuff). The first is what more people are used to, but the latter *is* more serviceable.

**Comment by davidben** [comment]
It's equivalent here. It's just a question of whether FOO_R_REASON has ERR_LIB_FOO ORd in. But, anyway, I was just answering why a module code. As things stand right now, you need the module code specified.

**Comment by levitte** [comment]
@richsalz, it's valuable to the users (you know, the programmers who work with our API) to get some knowledge from where an error in the queue came. You cannot ask each to download our source and start digging every time they get a stack of errors...

**Comment by levitte** [comment]
Well, yes, having the module OR'd in is exactly what I'm thinking. All it takes is reserving a bit field in the code for the module's number...

**Comment by t-j-h** [comment]
I think the main issue is to look at what application (or other library) code is doing with the error stack - i.e. what information is being used - and for what purpose and see what other approaches could be used.

This is entirely separate from the context of just printing out the error stack as diagnostic information - to me that is useful - but there are many ways of handling that - what counts is where do we have code that uses the information to do something other than printing it out. 

i.e. review the usage of ERR_peek_error and see how it is being picked up.

Typically (in my experience) there is basically a conversion of error stack information into something a little more meaningful to the end user.

OpenSSL itself have absolutely minimal usage of the error stack internally - once in ssl/ssl_rsa.c - which doesn't care what the actual error is, once in ssl/ssl_lib.c where it is simply distinguishing between libc errors and ssl errors in SSL_get_error, and once in crypto/asn1/a_d2i_fp.c where it is checking for invalid ASN1 bounds - so the library is not reflected of the actual usage.

e.g. just looking at the apache web server

https://github.com/apache/httpd/blob/trunk/modules/ssl/ssl_engine_init.c
https://github.com/apache/httpd/blob/trunk/modules/ssl/ssl_engine_io.c
https://github.com/apache/httpd/blob/trunk/modules/ssl/ssl_engine_log.c

Some of that sort of usage can probably be handled rather differently - but tackling how the error stack is being used in libraries and applications provides the context to keep in mind if restructuring.

**Comment by davidben** [comment]
Now that the next release is going to be an API break (?), time to drop the function codes? We've been running without them for some time. Having them available means people are likely to condition on them programmatically, which results in code that breaks whenever OpenSSL moves code around.

**Comment by levitte** [comment]
We're trying to aim for no breakage in the next release.  However, it should be possible to have a transition, i.e. current code is left as is, but a new variant is created that does away with certain data that's regarded as unnecessary.  Internally, they could simply get the same constant value in the function part of the number...

**Comment by davidben** [comment]
We just make `ERR_GET_FUNC` return zero. `ERR_func_error_string` returns the string `OPENSSL_internal`. That avoids breaking anyone trying to print the function codes as a string somewhere.

We intentionally leave `*_F_*` constants undefined. If your code references those, that means you were doing something like (from `apps/rsa.c`):
```
            while ((err = ERR_peek_error()) != 0 &&
                   ERR_GET_LIB(err) == ERR_LIB_RSA &&
                   ERR_GET_FUNC(err) == RSA_F_RSA_CHECK_KEY_EX &&
                   ERR_GET_REASON(err) != ERR_R_MALLOC_FAILURE) {
```
That kind of thing fundamentally can't work across OpenSSL versions as you move code around, so you want to break those so people can remove the `ERR_GET_FUNC` conditions.

**Comment by mspncp** [comment]
@davidben do you consider this issue completely resolved by #9058 or is there anything outstanding?
